runtime/commonMain/src/kotlinx/io/Streams.kt

```diff
@@ -1,49 +1,252 @@
-)
+
 package kotlinx.io
 
 import kotlinx.serialization.*
 
-@Deprecated(message = message, level = DeprecationLevel.ERROR)
-expect open class IOException: Exception {
-    constructor()
-    constructor(message: String)
-}
-
 @InternalSerializationApi
-expect abstract class InputStream {
-    open fun available(): Int
-    open fun close()
+abstract class Input {
     abstract fun read(): Int
-    open fun read(b: ByteArray): Int
-    open fun read(b: ByteArray, offset: Int, len: Int): Int
-    open fun skip(n: Long): Long
+    abstract fun read(b: ByteArray, offset: Int, length: Int): Int
+    abstract fun readString(length: Int): String
+    abstract fun readVarint32(): Int
+    abstract fun readVarint64(eofAllowed: Boolean): Long
 }
 
 @InternalSerializationApi
-expect class ByteArrayInputStream(buf: ByteArray): InputStream {
-    override fun read(): Int
+class ByteArrayInput(private var array: ByteArray) : Input() {
+
+    private var position: Int = 0
+
+    override fun read(): Int {
+        return if (position < array.size) array[position++].toInt() and 0xFF else -1
+    }
+
+    override fun read(b: ByteArray, offset: Int, length: Int): Int {
+        // avoid int overflow
+        if (offset < 0 || offset > b.size || length < 0
+            || length > b.size - offset
+        ) {
+            throw IndexOutOfBoundsException()
+        }
+        // Are there any bytes available?
+        if (this.position >= array.size) {
+            return -1
+        }
+        if (length == 0) {
+            return 0
+        }
+
+        val copied = if (this.array.size - position < length) this.array.size - position else length
+        array.copyInto(destination = b, destinationOffset = offset, startIndex = position, endIndex = position + copied)
+        position += copied
+        return copied
+    }
+
+    override fun readString(length: Int): String {
+        val result = array.decodeToString(position, position + length)
+        position += length
+        return result
+    }
+
+    override fun readVarint32(): Int {
+        if (position == array.size) {
+            error("Unexpected EOF")
+        }
+
+        // Fast-path: single and two byte values
+        var currentPosition = position
+        var result = array[currentPosition++].toInt()
+        if (result >= 0) {
+            position  = currentPosition
+            return result
+        } else if (array.size - position > 1) {
+            result = result xor (array[currentPosition++].toInt() shl 7)
+            if (result < 0) {
+                position = currentPosition
+                return result xor (0.inv() shl 7)
+            }
+        }
+
+        return readVarint32SlowPath()
+    }
+
+    override fun readVarint64(eofAllowed: Boolean): Long {
+        if (position == array.size) {
+            if (eofAllowed) return -1
+            else error("Unexpected EOF")
+        }
+
+        // Fast-path: single and two byte values
```

Comment:
Are there any text commentaries explaining this algorithm?

Reply:
Added small explanation